From 024c793b3b237ed4d9a72cd9b3f28ad51c4672f2 Mon Sep 17 00:00:00 2001
From: Igor Ulitin <iulitin@techinline.com>
Date: Wed, 17 Jun 2020 11:23:03 +0400
Subject: [PATCH] save any address for udp

---
 p2p/base/stunport.cc | 4 ----
 1 file changed, 4 deletions(-)

diff --git a/p2p/base/stunport.cc b/p2p/base/stunport.cc
index 727312329e..e59471e7dd 100644
--- a/p2p/base/stunport.cc
+++ b/p2p/base/stunport.cc
@@ -349,10 +349,6 @@ void UDPPort::OnLocalAddressReady(rtc::AsyncPacketSocket* socket,
   // applications which absolutely requires a HOST candidate.
   rtc::SocketAddress addr = address;
 
-  // If MaybeSetDefaultLocalAddress fails, we keep the "any" IP so that at
-  // least the port is listening.
-  MaybeSetDefaultLocalAddress(&addr);
-
   AddAddress(addr, addr, rtc::SocketAddress(), UDP_PROTOCOL_NAME, "", "",
              LOCAL_PORT_TYPE, ICE_TYPE_PREFERENCE_HOST, 0, "", false);
   MaybePrepareStunCandidate();
-- 
2.22.0.windows.1

